app/src/main/java/xyz/wim/dentrealitytask/ui/map/MapFragment.kt

```diff
@@ -1,107 +1,139 @@
         fun newInstance() = MapFragment()
-    }
 
-    private var root: View? = null
+        private const val MAPVIEW_BUNDLE_KEY = "MapViewBundleKey"
+    }
 
     private var _binding: MapFragmentBinding? = null
 
-    // This property is only valid between onCreateView and
-// onDestroyView.
+    // This property is only valid between onCreateView and onDestroyView.
     private val binding get() = _binding!!
 
     private val vm: MapViewModel by viewModel()
 
+    private val mapChannel = Channel<GoogleMap>()
+
     override fun onCreateView(
         inflater: LayoutInflater, container: ViewGroup?,
         savedInstanceState: Bundle?
-    ): View? {
+    ): View {
         _binding = MapFragmentBinding.inflate(inflater, container, false)
-        val view = _binding!!.root
-        return view
+        lifecycleScope.launch {
+            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
+                displayCountryMarkers()
+            }
+        }
+        with(binding) {
+            mapView.onCreate(savedInstanceState)
+            mapView.getMapAsync {
+                lifecycleScope.launch {
+                    mapChannel.send(it)
+                }
+            }
+        }
+
+        return _binding!!.root
     }
 
     override fun onCreate(savedInstanceState: Bundle?) {
         vm.countries.observe(this) {
             Timber.d("Observing countries")
-            createCountryViews(it)
         }
         super.onCreate(savedInstanceState)
     }
 
-    override fun onDestroyView() {
-        super.onDestroyView()
-        _binding = null
-    }
+    private suspend fun displayCountryMarkers() {
+        val map = mapChannel.receive()
+        val countries = vm.countries.await()
 
-    private fun createCountryViews(countries: List<Country>) {
         for (country in countries) {
-            createCountryView(country)
+            val latlng = LatLng(country.latlng[0], country.latlng[1])
+            map.addMarker(
+                MarkerOptions()
+                    .position(latlng)
+                    .title(country.countryCode)
+            )
+        }
+        map.setOnMarkerClickListener { marker ->
+            Timber.d("Marker clicked: %s", marker.title)
+            activity?.let { activity ->
+                activity.supportFragmentManager.beginTransaction()
+                    .replace(R.id.container, DetailsFragment.newInstance(marker.title!!))
+                    .addToBackStack(null)
+                    .commit()
+            }
+            return@setOnMarkerClickListener true
+        }
+    }
+
+    override fun onSaveInstanceState(outState: Bundle) {
+        super.onSaveInstanceState(outState)
+        var mapViewBundle = outState.getBundle(MAPVIEW_BUNDLE_KEY)
+
+        if (mapViewBundle == null) {
+            mapViewBundle = Bundle()
+            outState.putBundle(MAPVIEW_BUNDLE_KEY, mapViewBundle)
         }
+        _binding?.mapView?.onSaveInstanceState(mapViewBundle)
```

Comment:
This works for activity lifecycle events (like app going to background etc) but state is lost when you rotate screen or return from details screen. State should be saved using a ViewModel.